ext/strscan/strscan.c

```diff
@@ -1266,6 +1268,55 @@ strscan_peep(VALUE self, VALUE vlen)
     return strscan_peek(self, vlen);
 }
 
+/*
+ * call-seq:
+ *   scan_integer
+ *
+ * Equivalent to #scan with a [+-]?\d+ pattern, and returns an Integer or nil.
+ *
+ * The scanned string must be encoded with an ASCII compatible encoding, otherwise
+ * Encoding::CompatibilityError will be raised.
+ */
+static VALUE
+strscan_scan_integer(VALUE self)
+{
+    char *ptr, *buffer;
+    long len = 0;
+    VALUE buffer_v, integer;
+    struct strscanner *p;
+
+    GET_SCANNER(self, p);
+    CLEAR_MATCH_STATUS(p);
+
+    rb_must_asciicompat(p->str);
+
+    ptr = CURPTR(p);
+
+    if (ptr[len] == '-' || ptr[len] == '+') {
```

Comment:
Is `ptr[len]` safe when `S_RESTLEN()` is `0`?

Reply:
I think so, because (for now?) Ruby strings are guaranteed to be NULL terminated. So if we're EOF, `ptr[len]` is `\0`.

Reply:
It seems that a `String` created by `rb_str_new_static()` may not be terminated by `\0`.

Reply:
Technically you can but the documentation clearly states:

> Identical to rb_str_new(), except it takes a C string literal.

So it assume it's a literal, hence is NUL terminated.

I can add the checks if you feel strongly about it, but it's a lot of busy code for something I think isn't supposed to happen.